### ISSUE TITLE
How to exclude variables from imputation

### ISSUE THREAD

**Issue by squalo78** [stefvanbuuren/mice#202]
I am using a mock dataset, with 3 variables, where  I would like to only impute the third variable: VISIT_1
         
ID RND     VISIT_1
1     CTR.1 CTR          NA
2     CTR.2 CTR          NA
3     CTR.3 CTR          NA
4     CTR.4 CTR          NA
27   CTR.27 CTR -2.87737921
28   CTR.28 CTR -0.52460587
29   CTR.29 CTR  0.50137659
30   CTR.30 CTR  0.20553050
31   CTR.31 CTR  0.30132580
32   CTR.32 CTR -0.17399664
33   CTR.33 CTR -0.08097743
34   CTR.34 CTR  0.28508992
35   CTR.35 CTR -1.27311629
36   CTR.36 CTR -0.09858188
37   CTR.37 CTR -0.62438059
271   TRT.4 TRT          NA
272   TRT.5 TRT          NA
273   TRT.6 TRT          NA
274   TRT.7 TRT          NA
275   TRT.8 TRT          NA
276   TRT.9 TRT          NA
277  TRT.10 TRT          NA
278  TRT.11 TRT          NA
279  TRT.12 TRT          NA
280  TRT.13 TRT          NA
250 CTR.250 CTR  0.07292614
251 CTR.251 CTR  1.21464834
252 CTR.252 CTR -0.74335959
253 CTR.253 CTR  0.47283082

I am using the code below:
mice(my.data, method = c("","","norm"), m=N.MI, 
                 predictorMatrix = matrix(c(0, 0, 0,
                                                            0, 0, 0,
                                                            0, 0, 0), nrow=3, byrow=T),
                 printFlag=FALSE)

but with the current version of the package (it was working 1 year ago) I get the following error:

Error in edit.setup(data, setup, ...) : 
  `mice` detected constant and/or collinear variables. No predictors were left after their removal.

I have tried adding also: remove.collinear = FALSE and remove.constant = FALSE, but I get the same error.

So, what I am doing wrong? Do I misspecify the predictorMatrix?

Thanks for any help for understanding and apologizes if I am doing an obvious mistake

Giuseppe

**Comment by stefvanbuuren** [comment]
All entries are zero, so you didn't select any predictors.

**Comment by squalo78** [comment]
I tried also with 
predictorMatrix = matrix(c(0, 0, 0,
                                           0, 0, 0,
                                           0, 0, 1), nrow=3, byrow=T)
and I get the exact same error.

How you would write the predictorMatrix for such an example?

Thanks for any help.

**Comment by squalo78** [comment]
Very sorry stefvan to bother you again, but it would really appreciate if you could please help me addressing my last comment. I don't have anymore 'a prediction matrix will all zero entries', but I still get the same error.

**Comment by stefvanbuuren** [comment]
The predictor matrix is incorrect. 

You seem to be a novice. Start by using the default, and take some time to read the documentation.

Hope this helps.

**Comment by squalo78** [comment]
Thank you Stefvan.
Unfortunately this is not helpful, because in the documentation I could not find an answer to my question.

On version 2.46, the same code (method = c("","","norm") and predictorMatrix = matrix(c(0, 0, 0, 0, 0, 0, 0, 0, 0) ) was working. I guess the function would understand that no imputation is needed for the first 2 variables, while for the third variable it would impute from the 'same variable' (given that the corresponding 'row' has only zeros).

Given that you already pointed out 2 times that 'the predictor matrix is incorrect', would you be so kind to advise on which would be the right one to specify?

I have spent the last 2 weeks trying to figure out a solution (for my piece of code which was working 1 year ago, but not anymore) before writing you.

Thanks for any additional help.

Best Wishes
Giuseppe

**Comment by stefvanbuuren** [comment]
> while for the third variable it would impute from the 'same variable'

I think this is not possible because the third variable cannot appear at both sides of the equation in the imputation model. `mice` specifies the diagonal of the predictor matrix as 0. The information for imputation should come from other variables, and cannot come from the missing variable itself.

I never expected users to specify a 1 at the diagonal and zero elsewhere, so perhaps I should add a check for this case.

**Comment by squalo78** [comment]
Thank you Stefvan, I very much appreciate your answer.
Let me explain the rationale for my question.
For longitudinal studies, it is often the case that we collect multiple measurement for the same patient. For example, let's imagine that we have a response Y, and we collect Y at VISIT1, VISIT2,  VISIT3 (e.g. at 6 months, 12 months and 18 months)
Now if there are missing data for VISIT3, it would be natural to impute those with VISIT2 and VISIT3 as predictors. Similarly, if I have missing data for VISIT2, I would use VISIT1 as predictor.
But when I have missing data for VISIT1, then I would like to impute those with the mean of VISIT1.
So basically you are saying this is not possible.
Thanks again!

**Comment by gerkovink** [comment]
Dear Giuseppe,

Of course it is possible to impute the mean for any variable, but this may not be a good idea from an inference point of view. 

I suggest you take a look at the [mice vignettes](http://www.gerkovink.com/miceVignettes/). Many scenarios are detailed there. On another note: for longitudinal data, a simple data manipulation into a broad data format would allow for a more precise specification of the true data relations in the predictor matrix.

All the best,

Gerko
